.github/workflows/ci.yml

```diff
@@ -22,10 +22,21 @@ jobs:
       uses: actions/checkout@v2
       with:
         submodules: true
-    - name: Build ${{ matrix.distro }} Docker image with ${{ matrix.compiler }}
+    - name: Prepare base image in Docker
+      run: docker build . -f utils/docker/Dockerfile.${{ matrix.distro }}
+                          --build-arg DAOS_DEPS_BUILD=no
+    - name: Build dependencies in Docker
+      run: docker build . -f utils/docker/Dockerfile.${{ matrix.distro }}
+                          --build-arg DEPS_JOBS=10 --build-arg DAOS_BUILD=no
+    - name: Build Docker image with ${{ matrix.compiler }}
       run: docker build . -f utils/docker/Dockerfile.${{ matrix.distro }}
                           --build-arg COMPILER=${{ matrix.compiler }}
-                          -t ${{ matrix.distro }}/${{ matrix.compiler }}/daos:${{ github.sha }}
+                          --build-arg DEPS_JOBS=10 --build-arg DAOS_JAVA_BUILD=no
+    - name: Build Java in Docker image with ${{ matrix.compiler }}
+      run: docker build . -f utils/docker/Dockerfile.${{ matrix.distro }}
+                          --build-arg COMPILER=${{ matrix.compiler }}
+                          --build-arg DEPS_JOBS=10
```

Comment:
I haven't experimented with it yet, but it seems that docker in GH Actions will re-use the image/layers within a job?  Because I don't think it does any re-use of them outside of job does it?  Weren't you using an action to stash the docker images into the local organisation cache because of this?

Reply:
This section is all run on the same VM in order (any repeated over axes in the matrix) so within a PR and a specific build then it'll re-use them yes.  Splitting them like this here mostly allows finer grained viewing of progress and the logs rather than anything else.

Yes, I was stashing and re-using the images at one point, but I haven't yet found a clean way of doing that effectively.

Reply:
> run on the same VM

Ah right.

> Yes, I was stashing and re-using the images at one point, but I haven't yet found a clean way of doing that effectively.

￼You didn't like the method that action you were using was using?  It does seem like some native per-organisation caching of docker images/layers (without having to resort to tricks like that action uses) would be useful and beneficial to GH even, in not having to do full image builds for every action run.

Reply:
my previous attempt worked, and did reuse the intermediate images, which probably made us less susceptible to network errors etc but there's a limited size of 5GB cache per repository, and in practice it's just as quick to rebuild the image as it is to access the cache.  I've also tried collapsing the matrix and running all variants of the compiler build in the same VM (so just use the matrix for distro, but just loop over compiler) and that's faster as you can reuse the dependencies, in fact that's when I discovered that COMPILER was being set before the dependencies were build, but not used until after which is what this PR fixes.

Reply:
Yeah.  I was wondering whether the cache stashing/unstashing was any faster than rebuilding.  Which is a pity and why some more docker-native per-organisation caching would be useful.